cmd/datasource-token-refresh/main.go

```diff
@@ -0,0 +1,190 @@
+		if err != nil {
+			level.Error(logger).Log("msg", fmt.Sprintf("couldn't send update datasource request to datasource id: %s", datasourceUID), "err", err)
+			continue
+		}
+		dsSucessfullyUpdated = append(dsSucessfullyUpdated, datasourceUID)
+	}
+	if len(dsSucessfullyUpdated) != 0 {
+		level.Info(logger).Log("msg", fmt.Sprintf("Updated grafana datasource ids: %s", dsSucessfullyUpdated))
+	}
+}
+
+// getOAuth2Token generates an OAuth token based if a JSON file is provided or it will use the default credentials.
+func getOAuth2Token(jsonFilePath string) (string, error) {
+	var err error
+	var token oauth2.TokenSource
+	if jsonFilePath == "" {
+		ctx := context.Background()
+		token, err = google.DefaultTokenSource(ctx, "https://www.googleapis.com/auth/monitoring.read")
+		if err != nil {
+			return "", err
+		}
+	} else {
+		jsonKey, err := os.ReadFile(jsonFilePath)
+		if err != nil {
+			return "", fmt.Errorf("failed to read json key file: %v", err)
+		}
+		token, err = google.JWTAccessTokenSourceWithScope([]byte(jsonKey), "https://www.googleapis.com/auth/monitoring.read")
+		if err != nil {
+			return "", fmt.Errorf("could not generate token: %v", err)
+		}
+	}
+	accessToken, err := token.Token()
+	if err != nil {
+		return "", err
+	}
+	return accessToken.AccessToken, nil
+}
+
+/*
+buildUpdateDataSourceRequest takes an existing datasource config and adds or modifies the Authorication header
+and updates it to make Grafana compatible with GMP. For refrerence this is an example of a Grafana datasource:
+
+	"url": "https://monitoring.googleapis.com/v1/projects/gpe-test-1/location/global/prometheus/",
+	"jsonData": {
+	    "httpHeaderName1": "X-Custom-Header"
+	    "httpHeaderName2": "Authorization",
+	    "httpMethod": "POST",
+	    "prometheusType": "Prometheus",
+	    "prometheusVersion": "2.40.0",
+	},
+	"secureJsonFields": {
+	    "httpHeaderValue1": "secure value",
+	    "httpHeaderValue2": "secure value",
+	},
+*/
+func buildUpdateDataSourceRequest(dataSource grafana.DataSource, token string) (*grafana.DataSource, error) {
+	var (
+		authorizationHeader = "Authorization"
+		// httpHeader* are the prefixes that is used to store the name and value of of a custom headers.
+		// check https://github.com/grafana/grafana/blob/148e1c1588e9f075b14b72eb87d5463ea5bbb253/pkg/services/datasources/models.go#L34C1-L34C1 for more info.
+		httpHeaderName  = "httpHeaderName"
+		httpHeaderValue = "httpHeaderValue"
+	)
+	if dataSource.Type != "prometheus" {
+		return nil, fmt.Errorf("datasource type is not prometheus")
+	}
+
+	if *prometheusEndpoint != "" {
+		dataSource.URL = *prometheusEndpoint
+	}
+
+	if dataSource.URL == "" {
+		return nil, fmt.Errorf("prometheus-endpoint is not set")
+	}
+
+	jsonData := dataSource.JSONData
+	// Miscellaneous changes to make Grafana more compatible with GMP
+	jsonData["httpMethod"] = http.MethodGet
+	jsonData["prometheusType"] = "Prometheus"
+	jsonData["prometheusVersion"] = "2.40.0"
```

Comment:
Should we change this so that it only updates to 2.40.0 if blank or <2.40.0? I could see users wanting to make it higher for future reasons